pkg/server/plugin/keymanager/awskms/client_fake.go

```diff
@@ -0,0 +1,315 @@
+	if k.signErr != nil {
+		return nil, k.signErr
+	}
+
+	k.mu.RLock()
+	defer k.mu.RUnlock()
+	_, err := k.getKeyEntry(*input.KeyId)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO: do actual signing
+	return &kms.SignOutput{Signature: input.Message}, nil
+}
+
+func (k *kmsClientFake) CreateAlias(ctx context.Context, input *kms.CreateAliasInput, opts ...func(*kms.Options)) (*kms.CreateAliasOutput, error) {
+	if k.createAliasErr != nil {
+		return nil, k.createAliasErr
+	}
+
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	keyEntry, err := k.getKeyEntry(*input.TargetKeyId)
+	if err != nil {
+		return nil, err
+	}
+	k.aliases[*input.AliasName] = keyEntry
+
+	return &kms.CreateAliasOutput{}, nil
+}
+
+func (k *kmsClientFake) UpdateAlias(ctw context.Context, input *kms.UpdateAliasInput, opts ...func(*kms.Options)) (*kms.UpdateAliasOutput, error) {
+	if k.updateAliasErr != nil {
+		return nil, k.updateAliasErr
+	}
+
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	//TODO: review logic
+	keyEntry, err := k.getKeyEntry(*input.TargetKeyId)
+	if err != nil {
+		return nil, err
+	}
+	k.aliases[*input.AliasName] = keyEntry
+
+	return &kms.UpdateAliasOutput{}, nil
+}
+
+func (k *kmsClientFake) setEntries(entries []fakeKeyEntry) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	if entries == nil {
+		return
+	}
+	for _, e := range entries {
+		if e.KeyID != nil {
+			k.keyEntries[*e.KeyID] = e
+		}
+		if e.AliasName != nil {
+			k.aliases[*e.AliasName] = e
+		}
+	}
+}
+
+func (k *kmsClientFake) getKeyEntry(idOrAlias string) (fakeKeyEntry, error) {
+	keyEntry, ok := k.aliases[idOrAlias]
+	if ok {
+		return keyEntry, nil
+	}
+
+	keyEntry, ok = k.keyEntries[idOrAlias]
+	if ok {
+		return keyEntry, nil
+	}
+
+	return fakeKeyEntry{}, errors.New("no such key")
+}
+
+func (k *kmsClientFake) setCreateKeyErr(fakeError string) {
+	if fakeError != "" {
+		k.createKeyErr = errors.New(fakeError)
+	}
+}
+func (k *kmsClientFake) setDescribeKeyErr(fakeError string) {
+	if fakeError != "" {
+		k.describeKeyErr = errors.New(fakeError)
+	}
+}
+
+func (k *kmsClientFake) setgetPublicKeyErr(fakeError string) {
+	if fakeError != "" {
+		k.getPublicKeyErr = errors.New(fakeError)
+	}
+}
+
+func (k *kmsClientFake) setListAliasesErr(fakeError string) {
+	if fakeError != "" {
+		k.listAliasesErr = errors.New(fakeError)
+	}
+}
+
+func (k *kmsClientFake) setCreateAliasesErr(fakeError string) {
+	if fakeError != "" {
+		k.createAliasErr = errors.New(fakeError)
+	}
+}
+
+func (k *kmsClientFake) setUpdateAliasesErr(fakeError string) {
+	if fakeError != "" {
+		k.updateAliasErr = errors.New(fakeError)
+	}
+}
+
+func (k *kmsClientFake) setScheduleKeyDeletionErr(fakeError error) {
+	if fakeError != nil {
+		k.scheduleKeyDeletionErr = fakeError
+	}
+}
+
+func (k *kmsClientFake) setSignDataErr(fakeError string) {
+	if fakeError != "" {
+		k.signErr = errors.New(fakeError)
+	}
+}
+
+func generateRSAKey(bits int) (*rsa.PrivateKey, *rsa.PublicKey, error) {
```

Comment:
Key generation can sometimes negatively impact unit test times, particularly for the large key sizes. If possible, can we return pre-generated keys here, possibly from the `testkey` package?

Reply:
Thanks for that suggestion. On that note, let me know what you think about signing https://github.com/spiffe/spire/blob/b09793eddea2c3bba4f2e8156ad45a96d2f8e9dc/pkg/server/plugin/keymanager/awskms/client_fake.go#L189 Whether we should implement actual signing in the fake client or not.